criu/files-reg.c

```diff
@@ -321,19 +321,46 @@ static int mkreg_ghost(char *path, GhostFileEntry *gfe, struct cr_img *img)
 	return ret;
 }
 
+static int mklnk_ghost(char *path, GhostFileEntry *gfe)
+{
+	if (!gfe->symlnk_target) {
+		pr_err("Ghost symlink target is NULL for %s. Image from old CRIU?\n", path);
+		return -1;
+	}
```

Comment:
so, you're going to break images back-compatibility?
also, shouldn't it be `gfe->has_symlnk_target`?

Reply:
No, I'm not going to break back-compatibility, and this check does not break compatibility because for old images we have next situation:

1. Let suppose that we have CRIU without these changes and use old images.
2. Also, let suppose, that we have opened fd to symlink with `O_PATH`
3. Then we do `criu dump`... all fine
4. Then we do `criu restore` and we will fail on restore. Because we have this check:
https://github.com/checkpoint-restore/criu/blob/43a7b34094ddda2a5f60dc4902a55a1d73f6f6ec/criu/files-reg.c#L1760
we will fail with the error, that file has bad mode, because with old criu + old image we will restore our ghost symlink as regular ghost file.

Now, let's take a look on case with CRIU + this patch + old image:
after `criu restore` we will also fail. :) But error message says us, that our image is bad and not contains symlink target.
Speaking strictly, this is incompatible change because we have changed behavior (error messages), but from another point of view in both cases, our `criu restore` will fail.

>also, shouldn't it be gfe->has_symlnk_target

As far as I know, for `optional string` fields and sub-messages there are no `has_` field. And we should simply check pointer.

Reply:
Fair enough :)